### ISSUE TITLE
PAID [$500]  Web - Workspace - After deleting a bank account, workspace options remain available initially

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#11512]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to URL https://staging.new.expensify.com/
2. Login with any account
3. Create a Workspace (if it has not been created before)
4. Go to Settings->Workspace->Connect bank account->Connect online with Plaid
5. Complete the bank account creation process
6. Verify that the workspace options are available
7. Go to Settings->Workspace->Connect bank account->Disconnect bank account
8. Ensure that the workspace options are no longer available
9.  Check the Workspace Setting, make sure all options are locked now

## Expected Result:
Workspace options become unavailable

## Actual Result:
Workspace options remain available initially after disconnect bank account

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- IOS
- Android
- mWeb
- Desktop

**Version Number:** 1.11.2.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/193424020-50f4ebed-d4d0-4197-bbd2-db5c3243c9f9.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by danieldoglas** [comment]
This is indeed something that should be automatically updated. Checked and can be exported.

**Comment by ItsJustTomDev** [comment]
So far from what I can see is that when you disconnect your bank account, you still have access to the "issue cards" page.

https://github.com/Expensify/App/blob/18fcaf43ae69f6a9e445f586b3f90f8d582eff18/src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.js#L12

This function gets called when you disconnect your bank, so what does it do? Well it tells the API (backend) to update that this user with this and this ID has disconnect there bank account, but then there is no check of that in this page: https://github.com/Expensify/App/blob/18fcaf43ae69f6a9e445f586b3f90f8d582eff18/src/pages/workspace/card/WorkspaceCardPage.js#L23

There is no check that checks that there is a bank account connect to this user, so it will always show that the cards are ready.
So how would we fix this? There is actually already some kind of implementation that fixes this issue.

in the WorkspaceBankAccountPage is a fix, that checks if there is a bank account yes or the no.
https://github.com/Expensify/App/blob/18fcaf43ae69f6a9e445f586b3f90f8d582eff18/src/pages/workspace/WorkspaceBankAccountPage.js#L54

Here on component mount, it checks if there is a bank account connected,
This is the function its calling: https://github.com/Expensify/App/blob/18fcaf43ae69f6a9e445f586b3f90f8d582eff18/src/pages/workspace/WorkspaceBankAccountPage.js#L66

and this function is basically just checking with the backend if this user has a bank account connected.
So the fix is very simple, we just implement getShouldShowPage function into the cards page, we check it on mount, and if there is no open bank, we either show a different component, or we disable everything.

**Comment by mananjadhav** [comment]
@ItsJustTomDev Thanks for proposal. I am not sure I agree with the `getShouldShowPage` implementation, because we already have `WorkspacePageWithSections` page with multiple checks.

@kbecciv Can you confirm if its happening only on the Cards page? or other pages too?

**Comment by kbecciv** [comment]
@mananjadhav Its happening on other pages as well.

**Comment by ItsJustTomDev** [comment]
@mananjadhav @kbecciv, alright guys, I have not seen the `WorkspacePageWithSections` 😅
I will setup a local dev environment  up tomorrow so I can debug this issue, there should be something wrong with fetching the data.

**Comment by laurenreidexpensify** [comment]
@ItsJustTomDev are you still following up on this? Thanks

**Comment by bernhardoj** [comment]
Hi, I'm trying to test this issue on my end, but I stuck at Connect Bank Account step. I choose Connect online with Plaid and after its success, it asks me to fill Routing and Account number. I fill the account number with random number and routing number with the number I found on google and I can pass this. On the second step, it asks to fill more information that I also just fill it randomly. However, when I click Save & Continue, it gives me an error which can be seen on the image below
![Untitled](https://user-images.githubusercontent.com/50919443/196044289-152028ad-d935-439e-bb99-6c29ef713ffa.png)

**Comment by mananjadhav** [comment]
@laurenreidexpensify I think we should double this

**Comment by bernhardoj** [comment]
### **Proposal**
As @ItsJustTomDev mentioned, when we disconnect bank account, `resetFreePlanBankAccount` is called. Inside of it, we merge the `REIMBURSEMENT_ACCOUNT` with new `achData`. However, we forget to merge the "state" property (to reset it). So, we can just add **state** with value of **BankAccount.STATE.DELETED**.

```diff
function resetFreePlanBankAccount() {
    const bankAccountID = lodashGet(store.getReimbursementAccountInSetup(), 'bankAccountID');
    if (!bankAccountID) {
        throw new Error('Missing bankAccountID when attempting to reset free plan bank account');
    }
    if (!store.getCredentials() || !store.getCredentials().login) {
        throw new Error('Missing credentials when attempting to reset free plan bank account');
    }

    const achData = {
        useOnfido: true,
        policyID: '',
        isInSetup: true,
        domainLimit: 0,
        currentStep: CONST.BANK_ACCOUNT.STEP.BANK_ACCOUNT,
+        state: BankAccount.STATE.DELETED,
    };

    API.write('RestartBankAccountSetup',
        {
            bankAccountID,
            ownerEmail: store.getCredentials().login,
        },
        {
            optimisticData: [{
                onyxMethod: 'merge',
                key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
                value: {achData, shouldShowResetModal: false},
            },
            {
                onyxMethod: 'set',
                key: ONYXKEYS.REIMBURSEMENT_ACCOUNT_DRAFT,
                value: null,
            }],
        });

    Navigation.navigate(ROUTES.getBankAccountRoute());
}
```

**Comment by mananjadhav** [comment]
I think @bernhardoj's proposal will work. I am good with it.

@puneetlath All yours.

🎀 👀 🎀 
C+ reviewed

**Comment by puneetlath** [comment]
Nice and simple! I like it.

**Comment by bernhardoj** [comment]
I have applied in Upwork and expect the PR to be delivered today or tomorrow.

**Comment by bernhardoj** [comment]
I have opened the PR here https://github.com/Expensify/App/pull/12279. Kindly review it. Thanks!

(Ignore my first PR, I put the wrong issue link. Sorry for the mistake)

**Comment by mananjadhav** [comment]
@kbecciv Can you help me with some form data for Bank Account creation?

I am stuck at this step:

<img width="387" alt="image" src="https://user-images.githubusercontent.com/3069065/199036364-6accd0bc-9476-4754-a73e-dbbc92c34154.png">

**Comment by mananjadhav** [comment]
@kbecciv Quick bump. Can you help me with the QA steps for this one?

**Comment by kbecciv** [comment]
Sorry, I missed you tread earlier.
Please use the steps for adding fully verified withdrawal account credentials:
1. Navigate to the add bank account modal in Workspace settings
2. Verify the Add bank account modal is displayed with 2 options to add bank accounts (Log in and Manual)
3. Select  plaid modal option
4. Select Fidelity bank
5. Enter the test credentials: Username: user_good | Password: pass_good
6. Choose the savings account "Plaid Saving11122XXXXXX111"
7. In the company information fill out the fields with the required test credentials
CompanyName: Alberta Bobbeth Charleson
CompanyTaxID: 123456789
First Name: Alberta
Last Name: Charleson 
8. Verify you're able to continue to the personal information step.
9. Fill out the fields with the required test credentials above
10. Verify you're able to continue.
11. In the beneficial owners step toggle the last 2 checkboxes (terms and conditions and true information)
12. Confirm you're able to continue and reach the validate step.
13. Verify the bank account was added to the workspace

**Comment by mananjadhav** [comment]
Okay thanks for the comment @kbecciv. Will I need to upload files with onfido with the given credentials?

**Comment by puneetlath** [comment]
@mananjadhav I don't think you should need to. Did you try it?

**Comment by mananjadhav** [comment]
@puneetlath I'll be up in 2 hours, will test and confirm.

**Comment by mananjadhav** [comment]
PR Review is done. For some reason the flow didn't work for my usual account. I logged in from another account and the flow with the test information worked. Thanks @kbecciv for the help here.

**Comment by puneetlath** [comment]
@mananjadhav could you help me with identifying the PR that caused the bug?

**Comment by mananjadhav** [comment]
@puneetlath I tried checking the PR, but the file has gone through multiple refactors. I'll try to take another stab at it.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by laurenreidexpensify** [comment]
Everyone is hired in Upwork

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by laurenreidexpensify** [comment]
Oh this was deployed to prod 8 days ago but for some reason the automation failed to flag it https://github.com/Expensify/App/pull/12279. Updating now.

**Comment by laurenreidexpensify** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist will need to be completed before the issue can be closed:

- [ ] [@puneetlath @mananjadhav] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@puneetlath @mananjadhav] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. 
- [ ] [@@puneetlath @mananjadhav] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@laurenreidexpensify] A [regression test](https://expensify.testrail.io/index.php?/dashboard) has been added or updated so that the same bug will not reach production again. Link to the GH issue for creating the test here:
- [ ] [@laurenreidexpensify] Applicable payments have been made. This potentially includes the issue reporter, the contributor that fixed the issue, and/or the contributor+ that helped on the issue.

**Comment by puneetlath** [comment]
@mananjadhav any luck with the offending PR?

**Comment by mananjadhav** [comment]
Well @puneetlath not so much. I traced the changes to the following PRs:

1. `src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.js` file to this https://github.com/Expensify/App/pull/6230, where we added the file, but I am not sure if we had the same object passed as this PR is like a refactor.


2.  `src/pages/workspace/WorkspacePageWithSections.js` changes to PR https://github.com/Expensify/App/pull/9138, where we changed the hasVBA check to `BankAccount.STATE.OPEN`

To further find the offending PR, I'll have to dig further to get the context of the changes. Do one of these PRs help?
